<launch>

  <arg name="ticks_per_meter" value="8484" />
  <arg name="wheel_separation" value="0.22" />

  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find rover)/description/rover_reversed.urdf.xml"/>

  <!-- Publish the robot state -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher">
    <param name="publish_frequency" value="10.0"/>
  </node>

  <!-- Provide simulated control of the robot joint angles -->
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
    <param name="use_gui" value="False" />
    <param name="rate" value="10.0"/>
  </node>

  <node pkg="tf" type="static_transform_publisher" name="imu_link_broadcaster" args="0 0 0 0 0 0 imu_base imu_link 100" />
  
  <node name="odom_publisher" pkg="diff_drive" type="diff_drive_odometry" output="screen">
    <rosparam subst_value="true">
      ticks_per_meter: $(arg ticks_per_meter)
      wheel_separation: $(arg wheel_separation)
      encoder_min: -32768
      encoder_max: 32768
      publish_tf: true
    </rosparam>
  </node>

  <node name="rover_manual_driver" pkg="rover" type="rover_manual_driver.py" output="screen">
    <param name="~reverse_motor_direction" value="true" />
    <remap from="~lwheel_ticks" to="lwheel_ticks" />
    <remap from="~rwheel_ticks" to="rwheel_ticks" />
    <remap from="~lwheel_speed" to="lwheel_speed" />
    <remap from="~rwheel_speed" to="rwheel_speed" />
    <remap from="~imu_link" to="imu/data_raw" />
  </node>
  <!--
  <node pkg="imu_calib" type="apply_calib" name="apply_calib" output="screen" respawn="false">
    <param name="calib_file" value="$(find rover)/config/imu_calib.yaml" />
    <param name="calibrate_gyros" value="true" />
    <remap from="corrected" to="imu/data_raw" />
  </node>
  -->
  <!-- Filter and fuse raw imu data -->
  <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_madgwick" output="screen" respawn="false">
    <param name="fixed_frame" value="odom" />
    <param name="use_mag" value="false" />
    <param name="publish_tf" value="true" />
    <param name="use_magnetic_field_msg" value="false" />
    <param name="world_frame" value="enu" />
    <!--<param name="remove_gravity_vector" value="true" />-->
    <!--<param name="orientation_stddev" value="0.05" />-->
    <remap from="imu/data" to="imu_link" />
  </node>
  <!--
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization_odom" clear_params="true"> 
    <rosparam command="load" file="$(find rover)/config/robot_localization.yaml" />
  </node>
  -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find rover)/config/imu.rviz" />

</launch>
